rust/datafusion/src/sql/utils.rs

```diff
@@ -0,0 +1,305 @@
+                find_exprs_in_expr(nested_expr.as_ref(), test_fn)
+            }
+            Expr::IsNull(nested_expr) => {
+                find_exprs_in_expr(nested_expr.as_ref(), test_fn)
+            }
+            Expr::Not(nested_expr) => find_exprs_in_expr(nested_expr.as_ref(), test_fn),
+            Expr::ScalarFunction { args, .. } => find_exprs_in_exprs(&args, test_fn),
+            Expr::ScalarUDF { args, .. } => find_exprs_in_exprs(&args, test_fn),
+            Expr::Sort {
+                expr: nested_expr, ..
+            } => find_exprs_in_expr(nested_expr.as_ref(), test_fn),
+
+            // These expressions don't nest other expressions.
+            Expr::Column(_)
+            | Expr::Literal(_)
+            | Expr::ScalarVariable(_)
+            | Expr::Wildcard => vec![],
+        }
+    };
+
+    matched_exprs.into_iter().fold(vec![], |mut acc, expr| {
+        if !acc.contains(&expr) {
+            acc.push(expr)
+        }
+
+        acc
+    })
+}
+
+/// Convert any `Expr` to an `Expr::Column`.
+pub(crate) fn expr_as_column_expr(expr: &Expr, plan: &LogicalPlan) -> Result<Expr> {
+    match expr {
+        Expr::Column(_) => Ok(expr.clone()),
+        _ => Ok(Expr::Column(expr.name(&plan.schema())?)),
+    }
+}
+
+/// Rebuilds an `Expr` as a projection on top of a collection of `Expr`'s.
+///
+/// For example, the expression `a + b < 1` would require, as input, the 2
+/// individual columns, `a` and `b`. But, if the base expressions already
+/// contain the `a + b` result, then that may be used in lieu of the `a` and
+/// `b` columns.
+///
+/// This is useful in the context of a query like:
+///
+/// SELECT a + b < 1 ... GROUP BY a + b
+///
+/// where post-aggregation, `a + b` need not be a projection against the
+/// individual columns `a` and `b`, but rather it is a projection against the
+/// `a + b` found in the GROUP BY.
+pub(crate) fn rebase_expr(
+    expr: &Expr,
+    base_exprs: &Vec<Expr>,
+    plan: &LogicalPlan,
+) -> Result<Expr> {
+    clone_with_replacement(expr, &|nested_expr| {
+        if base_exprs.contains(nested_expr) {
+            Ok(Some(expr_as_column_expr(nested_expr, plan)?))
+        } else {
+            Ok(None)
+        }
+    })
+}
+
+/// Determines if the set of `Expr`'s are a valid projection on the input
+/// `Expr::Column`'s.
+pub(crate) fn can_columns_satisfy_exprs(
+    columns: &Vec<Expr>,
+    exprs: &Vec<Expr>,
+) -> Result<bool> {
+    columns.iter().try_for_each(|c| match c {
+        Expr::Column(_) => Ok(()),
+        _ => Err(DataFusionError::Internal(
+            "Expr::Column are required".to_string(),
+        )),
+    })?;
+
+    Ok(find_column_exprs(exprs).iter().all(|c| columns.contains(c)))
+}
+
+/// Returns a cloned `Expr`, but any of the `Expr`'s in the tree may be
+/// replaced/customized by the replacement function.
+fn clone_with_replacement<F>(expr: &Expr, replacement_fn: &F) -> Result<Expr>
+where
+    F: Fn(&Expr) -> Result<Option<Expr>>,
+{
```

Comment:
Why result<option ?

Reply:
I should add some notes in the docstring on this. 

The `Result` allows `replacement_fn()` to be fallible, as is the case with `rebase_expr()`'s closure's use of `expr_as_column_expr()`. 

If `replacement_fn()` returns:

* `Err(_)`: The error is returned by `clone_with_replacement()`. 
* `Ok(Some(replacement))`: A replacement was provided, directing `clone_with_replacement()` to use the replacement. In this case, cloning will not continue down this particular branch of expressions.
* `Ok(None)`: A replacement was not provided, and cloning continues down this particular branch of expressions. 

Does that make sense?

Reply:
Updated in 7e61b200bab7411524b648c38a2a7e107fe1c83e.